File: solution-python/008_string_to_integer_atoi.py
"""
	008. String to Integer (atoi)
	@name: li jin
	@date: Feb 20, 2016
	@link: https://leetcode.com/problems/string-to-integer-(atoi)/
	@time: 64 ms
"""

class Solution:
    def myAtoi(self, str):
        str = str.lstrip()
        res, n, tag = 0, len(str), False
        if n:
            tag = str[0] == "-"
            pos = 1 if str[0] == "+" or str[0] == "-" else 0
            while pos < n:
                digit = str[pos]
                pos += 1
                if "0" <= digit <= "9":
                    res = res * 10 + int(digit)
                else:
                    break
        return max(-res, -2147483648) if tag else min(res, 2147483647)

File: solution-cpp/008_string_to_integer_atoi.cpp
/*
	008. String to Integer (atoi)
	@name: li jin
	@date: Feb 20, 2016
	@link: https://leetcode.com/problems/string-to-integer-(atoi)/
	@time: 8 ms
*/

class Solution {
public:
    int myAtoi(string str) {
        int n = str.length(), pos = 0;
        bool tag = false;
        double res = 0.0;
        if (n) {
            while (str[pos] == ' ') pos++;
            tag = str[pos] == '-';
            if(tag || str[pos] == '+') pos++;
            while (pos < str.length()) {
                char digit = str[pos++];
                if (digit < '0' || digit > '9') break;
                res = res*10 + digit - '0';
                if (res >= 2147483648)
                    return tag ? -2147483648 : 2147483647;
            }
        }
        return int(tag ? -res : res);
    }
};
